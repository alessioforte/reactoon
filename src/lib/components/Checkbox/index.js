import React, { useState } from "react";
import PropTypes from "prop-types";
import styled, { withTheme, ThemeProvider } from "styled-components";
import Theme, { getContrastYIQ } from "../../theme";
import Icon from "../Icon";

const Checkbox = ({ checked, label, onChange, theme }) => {
  const [value, setValue] = useState(checked || false)

  const handleOnChande = () => {
    setValue(!value)
    let data = { checked: !value, label }
    onChange(data)
  }

  return (
    <ThemeProvider theme={theme}>
      <Box onClick={handleOnChande}>
        <Check check={value}>
          <Icon
            name="checkmark"
            size="9px"
            color={theme.colors[getContrastYIQ(theme.colors.primary)]}
          />
        </Check>
        {label && <Label>{label}</Label>}
      </Box>
    </ThemeProvider>
  );
};

Checkbox.propTypes = {
  label: PropTypes.string,
  checked: PropTypes.bool,
  onChange: PropTypes.func
};

Checkbox.defaultProps = {
  checked: false,
  label: null,
  onChange: () => {},
  theme: Theme.styles
};

export default withTheme(Checkbox);

const Box = styled.div`
  display: inline-flex;
  align-items: center;
  cursor: pointer;
`;
const Check = styled.div`
  flex-shrink: 0;
  width: 16px;
  height: 16px;
  border-radius: 3px;
  background: ${props =>
    props.check ? props.theme.colors.primary : props.theme.colors.background};
  cursor: pointer;
  transition: all 0.3s ease;
  display: flex;
  align-items: center;
  justify-content: center;
  &:hover {
    opacity: 0.8;
  }
  & svg {
    opacity: ${props => (props.check ? 1 : 0)};
  }
`;
const Label = styled.div`
  display: inline-block;
  margin-left: 5px;
`;
