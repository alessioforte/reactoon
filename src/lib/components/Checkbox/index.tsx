import React, { useState, FC } from 'react';
import styled, { withTheme, ThemeProvider } from 'styled-components';
import { getContrastYIQ, styles } from '../../theme';
import Icon from '../Icon';
import { focus } from '../Styled/css';

interface Props {
  label?: string;
  name?: string;
  checked?: boolean;
  onChange?: (value: any) => void;
  theme?: any;
}

const Checkbox: FC<Props> = ({
  label,
  name = '',
  checked = false,
  onChange = value => console.log(value),
  theme = styles
}) => {
  const [value, setValue] = useState(checked || false);

  const handleOnClick = e => {
    e.preventDefault();
    const newValue = !value;
    setValue(newValue);

    const data = { value: newValue, checked: newValue, label, name };
    onChange(data);
  };

  return (
    <ThemeProvider theme={theme}>
      <Wrapper onClick={handleOnClick}>
        <input type='checkbox' name={name} value={value.toString()} />
        <Check
          role='checkbox'
          name={name}
          value={value}
          aria-label={label}
          aria-checked={value}
          tabIndex={0}
        >
          <Icon
            name='checkmark'
            size='9px'
            color={theme.colors[getContrastYIQ(theme.colors.primary)]}
          />
        </Check>
        {label && <Label>{label}</Label>}
      </Wrapper>
    </ThemeProvider>
  );
};

export default withTheme(Checkbox);

const Wrapper = styled.label`
  display: inline-flex;
  width: fit-content;
  cursor: pointer;
  font-size: 1rem;
  & > input {
    position: absolute;
    opacity: 0;
    z-index: -1;
  }
`;
const Check = styled.span<{ name: string; value: boolean }>`
  flex-shrink: 0;
  width: 20px;
  height: 20px;
  border-radius: 3px;
  background: ${props =>
    props.value ? props.theme.colors.primary : props.theme.colors.background};
  cursor: pointer;
  transition: all 0.3s ease;
  display: flex;
  align-items: center;
  justify-content: center;
  &:hover {
    opacity: 0.8;
  }
  &:focus {
    ${focus}
  }
  & svg {
    position: absolute;
    opacity: ${props => (props.value ? 1 : 0)};
  }
`;
const Label = styled.span`
  display: inline-block;
  margin-left: 5px;
`;
