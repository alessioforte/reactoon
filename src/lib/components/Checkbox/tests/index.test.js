import React from 'react';
import { render } from '@testing-library/react';
import Checkbox from '../index';

describe('<Checkbox />', () => {
  it('should render and match the snapshot', () => {
    const {
      container: { firstChild }
    } = render(<Checkbox />);
    expect(firstChild).toMatchSnapshot();
  });
});
