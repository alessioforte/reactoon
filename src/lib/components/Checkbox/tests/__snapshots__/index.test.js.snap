// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> should render and match the snapshot 1`] = `
<div
  class="sc-htpNat jmEnvn"
>
  <div
    class="sc-bxivhb lijjMI"
  >
    <div
      class="sc-bwzfXH cusHsC"
    >
      <svg
        class="sc-bdVaJa htzKZE"
        color="#1a1a1a"
        height="9px"
        viewBox="0 0 712 512"
        x="0px"
        y="0px"
      >
        <path
          d="M256.2,512c-14.4,0-28.9-5.5-39.9-16.5L16.8,295.9c-22-22-22-57.8,0-79.8s57.8-22,79.8,0l159.6,159.6L615.4,16.5c22-22,57.8-22,79.8,0c22.1,22,22.1,57.8,0,79.8L296.1,495.5C285.1,506.5,270.7,512,256.2,512z"
        />
      </svg>
    </div>
  </div>
</div>
`;
